src/std/crypto/libcrypto.ss

```diff
@@ -653,4 +675,190 @@ END-C
+}
+
+
+static int ffi_BIO_read(BIO *b, ___SCMOBJ o, int dlen)
+{
+ int len;
+ if (dlen < 0) {
+  len = U8_LEN(o);
+ } else {
+  len = dlen;
+ }
+ return BIO_read(b, U8_DATA(o), len);
+ }
+
+END-C
+)
+
+
+(c-define-type BIO "BIO")
+(c-define-type BIO* (pointer BIO (BIO*) "ffi_release_BIO"))
+
+(c-define-type BIO_METHOD "BIO_METHOD")
+(c-define-type BIO_METHOD* (pointer BIO_METHOD (BIO_METHOD*)))
+
+
+(define-c-lambda BIO_new (BIO_METHOD*) BIO*)
+(define-c-lambda/const-pointer BIO_s_mem () BIO_METHOD*)
+
+(define-c-lambda BIO_new_mem_buf(scheme-object) BIO*
+  "ffi_BIO_new_mem_buf")
+
+(define-c-lambda BIO_read(BIO* scheme-object int) int
+  "ffi_BIO_read")
+
+;; OSSL Decoder
+
+(c-declare #<<END-C
+static ___SCMOBJ ffi_release_OSSL_DECODER_CTX (void *ptr)
+{
+ OSSL_DECODER_CTX_free((OSSL_DECODER_CTX*)ptr);
+ return ___FIX (___NO_ERR);
+ }
+
+static ___SCMOBJ ffi_release_OSSL_ENCODER_CTX (void *ptr)
+{
+ OSSL_ENCODER_CTX_free((OSSL_ENCODER_CTX*)ptr);
+ return ___FIX (___NO_ERR);
+}
+
+
+static EVP_PKEY* 
+ ffi_EVP_PKEY_new_from_bio(BIO *in,
+						 const char *input_type,
+						 const char *keytype, int selection)
+{
+EVP_PKEY *pkey = NULL;
+OSSL_DECODER_CTX *dctx;
+// const char *input_type = "PEM";   /* NULL for any format */
+//const char *structure = NULL; /* any structure */
+//const char *keytype = "RSA";  /* NULL for any key */
+// int selection = 0;
+
+/* OSSL_DECODER_CTX_new_for_pkey(EVP_PKEY **pkey,
+                              const char *input_type,
+                              const char *input_struct,
+                              const char *keytype, int selection,
+                              OSSL_LIB_CTX *libctx, const char *propquery);
+*/
+
+dctx = OSSL_DECODER_CTX_new_for_pkey(&pkey, input_type, NULL,
+                                     keytype,
+                                     selection,
+                                     NULL, NULL);
+if (dctx == NULL) {
+  perror("ffi_EVP_PKEY_new_from_bio: no suitable potential decoders found");
+	}
+if (OSSL_DECODER_from_bio(dctx, in)) {
+	       			      // success!
+
+  EVP_PKEY_up_ref(pkey);
+  OSSL_DECODER_CTX_free(dctx);
+  return pkey;			      
+} else {
+  
+  perror("ffi_EVP_PKEY_new_from_bio: Cannot decode bio");
+	}
+  OSSL_DECODER_CTX_free(dctx);
+  return pkey;
+}
+
+static int ffi_EVP_PKEY_print_private(BIO *out, const EVP_PKEY *pkey)
+{
+ return EVP_PKEY_print_private(out, pkey, 1, NULL);
+}
+
+
+
+static int ffi_EVP_PKEY_to_bio(EVP_PKEY *pkey,
+			       BIO *out,
+			       const char *format,
+			       const char *structure,                                     int selection)
+{
+ 
+ OSSL_ENCODER_CTX *ectx;
+
+ ectx = OSSL_ENCODER_CTX_new_for_pkey(pkey,
+                                     selection,
+                                     format, structure,
+                                     NULL);
+
+ if (ectx == NULL) {
+    /* error: no suitable potential encoders found */
+  perror("ffi_EVP_PKEY_to_bio: no suitable potential encoders found");
```

Comment:
please don't perror right and left.

Also, this code is broken, as it falls through; you should return and error.